### ISSUE TITLE
melt has memory leak if producer is xml

### ISSUE THREAD

**Issue by Adonis-Song** [mltframework/mlt#832]
Hi. 
My mlt version is v7.8.0. System version is Mac Big Sur 11.5.1 .

When I debug melt, I find producer would not be close if I pass a path of xml to melt.

This is my command:

```
./melt /Users/apple/test.xml
```

And this is xml:
```
<?xml version="1.0" encoding="utf-8"?>
<mlt LC_NUMERIC="zh_CN.UTF-8" version="7.8.0" root="/Users/apple/Documents/vizard/bpmlt/build/out/bin">
  <consumer f="mp4" movflags="+faststart" vcodec="libx264" crf="23" mlt_service="avformat" width="1920" height="1080" rescale="bilinear" preset="fast" acodec="aac" frame_rate_num="25" frame_rate_den="1" target="/Users/apple/Documents/mediaframework/profile-test/1/1-5-mlt.mp4"/>
  <producer id="producer0">
    <property name="resource">/Users/apple/Documents/mediaframework/profile-test/1/1.mp4</property>
    <property name="mlt_service">avformat</property>
  </producer>
  <playlist id="playlist0">
    <entry producer="producer0" in="00:00:00.000" out="00:01:48.990"/>
  </playlist>
  <tractor id="tractor0" in="" out="">
    <track producer="playlist0"/>
  </tractor>
</mlt>
```

![20221020-115027](https://user-images.githubusercontent.com/34226411/196852200-7171869b-694b-41ab-a182-319972190f58.jpeg)

When I debug this point, I find 'melt' would not be close ,because it's refs is two.

If I add these code to close it , melt whill be crashed. 

```
            int count = mlt_properties_ref_count( MLT_PRODUCER_PROPERTIES( melt ));
            while (count > 1)
                mlt_producer_close(melt);
            mlt_producer_close(melt);
```

So I donot know why the refs of `melt` is two and when I try to close it , the process of melt will be crashed .

**Comment by bmatherly** [comment]
```
            int count = mlt_properties_ref_count( MLT_PRODUCER_PROPERTIES( melt ));
            while (count > 1)
                mlt_producer_close(melt);
            mlt_producer_close(melt);
```

This while loop is an infinite loop because count is never changed.

**Comment by Adonis-Song** [comment]
o.o  
Sorry, I copy  wrong code. It should be as follow:
```
            while (mlt_properties_ref_count( MLT_PRODUCER_PROPERTIES( melt )) > 1)
                mlt_producer_close(melt);
            mlt_producer_close(melt);
```
I found I cannot add `consumer` tag  to xml. If add `consumer`,  memory will be leak

**Comment by Adonis-Song** [comment]
Maybe I find why.

REPODUCER：
1. A melt xml that contains  `consumer` tag and doesnot contain `profile` tag.
2. run cmd ` melt <path to xml>`
3. debug the process  and you will find `melt` producer cannot be free
![image](https://user-images.githubusercontent.com/34226411/201633962-023c8665-6e45-49e9-b558-c3201ebf18bc.png)

This is my solution to fix it.
When I add  `consumer` tag to xml, `melt` producer will has two ref, one is from itself, the other is from `consumer`, so we must free `consumer` here. My change is as follows:
```
        if ( ! profile->is_explicit )
        {
            mlt_profile_from_producer( profile, melt );
            
           //free the consumer
            mlt_consumer melt_consumer = MLT_CONSUMER( mlt_service_consumer( MLT_PRODUCER_SERVICE( melt ) ) );
            if ( melt_consumer )
            {
                mlt_service_identify(mlt_producer_service(melt));
                mlt_consumer_close(melt_consumer);
            }

            mlt_producer_close( melt );
            melt = mlt_factory_producer( profile, "melt", &argv[ 1 ] );
        }
```

However, it will be crash  because `melt_consumer` has been set as a property when mlt parses xml. Source code as follow:
https://github.com/mltframework/mlt/blob/d7d2e9b1552202b2fcd7cadca63a4b13677e553e/src/modules/xml/producer_xml.c#L2301-L2302

So I think I must remove this prop before  free `consumer`, So I make some change in `producer_melt.c` file.
My change is as follow: 
```
    if ( producer )
    {
        mlt_consumer consumer = mlt_properties_get_data( MLT_PRODUCER_PROPERTIES(producer), "consumer", NULL );
        if ( consumer )
        {
            mlt_consumer_connect( consumer, MLT_PRODUCER_SERVICE(prod) );
           
            // add these code to remove prop
            mlt_properties_inc_ref(mlt_consumer_properties(consumer));
            mlt_properties_clear(MLT_PRODUCER_PROPERTIES(producer), "consumer");
        }
    }
```
This is the postion of source code
https://github.com/mltframework/mlt/blob/d7d2e9b1552202b2fcd7cadca63a4b13677e553e/src/modules/core/producer_melt.c#L549-L554

But I don't know if it's correct, Anyone has any comments?

**Comment by Adonis-Song** [comment]
Hello , anyone has any  comments?
@bmatherly  Sorry for bothering you, Could you please help me to check if the modification is correct?

**Comment by ddennedy** [comment]
Your analysis is correct, but probably not your change. I cannot read you change because it is not in diff format, and I am on a phone in the airport. I cannot do anything on this for a couple of weeks. You must wait until I get around to it. Besides, this is low priority because you did not put a profile in your XML, which is not normal.

**Comment by bmatherly** [comment]
I am willing to analyze and test the proposed change. But I would also need to see the change in a diff format (or even better, a pull request) so that I can be sure I understand the proposed change.

**Comment by ddennedy** [comment]
You cannot clear the consumer property.

**Comment by Adonis-Song** [comment]
> Reference
OK, maybe I donot need to clear it.  just need to make it NULL

**Comment by ddennedy** [comment]
fixed in 769cf9d
